UserRelated/Main.py

```diff
@@ -0,0 +1,208 @@
+
+    return parser
+
+
+def add_interest_points_shingles(shing_set: set, args: argparse.Namespace):
+    """
+    Adds interest points shingles shingles to shing_set.
+    :param shing_set: the user preferences shingle set.
+    :param args: the given command-line arguments.
+    """
+    if args.waterfall: shing_set.add(PointTag.WATERFALL.value)
+    if args.birding: shing_set.add(PointTag.BIRDING.value)
+    if args.river: shing_set.add(PointTag.RIVER.value)
+    if args.cave: shing_set.add(PointTag.CAVE.value)
+    if args.lake: shing_set.add(PointTag.WATER.value)
+    if args.spring: shing_set.add(PointTag.SPRING.value)
+    if args.geo: shing_set.add(PointTag.GEOLOGIC.value)
+    if args.historic: shing_set.add(PointTag.HISTORIC.value)
+
+
+def add_length_shingle(shing_set: set, args: argparse.Namespace):
+    """
+    Adds the appropriate length shingle to the shingles set.
+    :param shing_set: the user preferences shingle set
+    :param args: the given command-line arguments.
+    """
+    if args.length == 1:
+        shing_set.add(TrackLength.SHORT.value)
+    elif args.length == 2:
+        shing_set.add(TrackLength.MEDIUM.value)
+    else:
+        shing_set.add(TrackLength.LONG.value)
+
+
+def add_difficulty_shingle(shing_set: set, args: argparse.Namespace):
+    """
+    Adds the appropriate difficulty shingle to the shingles set.
+    :param shing_set: the user preferences shingle set
+    :param args: the given command-line arguments.
+    """
+    if args.difficulty == 1:
+        shing_set.add(TrackDifficulty.EASY.value)
+    elif args.difficulty == 2:
+        shing_set.add(TrackDifficulty.INTERMEDIATE.value)
+    else:
+        shing_set.add(TrackDifficulty.DIFFICULT.value)
+
+
+def add_shape_shing(shing_set: set, args: argparse.Namespace):
+    """
+    Adds the appropriate shape shingle to the shingles set.
+    :param shing_set: the user preferences shingle set
+    :param args: the given command-line arguments.
+    """
+    shing_set.add(TrackShape.LOOP.value) if args.shape == 1 else shing_set.add(TrackShape.CURVE.value)
+
+
+# Todo: this is a code duplication of the slopes_poc branch, we should have a separate class for LSH-ing we can reuse.
+def get_min_hash(shingles: set) -> MinHash:
+    """
+    given a set of shingles, creates a MinHash object updated with those shingles.
+    :param shingles: a set of track shingles.
+    :return: a MinHash object updated with the given shingles.
+    """
+    track_min_hash = MinHash(num_perm=128)
+    for shin in shingles:
+        track_min_hash.update(str(shin).encode('utf-8'))
+    return track_min_hash
+
+
+def create_user_shingles(args: argparse.Namespace) -> set:
+    """
+    Creates a set of shingles defining the user's path preferences.
+    :param args: the given command-line arguments.
+    :return: a set of shingles as described.
+    """
+    shing = set()
+    add_interest_points_shingles(shing, args)
```

Comment:
if we support None then  in interest points, we should add a Not <featute> tag in case the user explicitly said he didn't want some feature. and add nothing if he inputted None.

Reply:
See my answer here: https://github.com/noystl/NeedleProject/pull/2/files/a90cf616c2e763ff1671db002b2ea9893407c533#r389257093
I also added what you wrote to the project's backlog.